bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go

```diff
@@ -137,8 +138,22 @@ type File struct {
 	// +optional
 	Encoding Encoding `json:"encoding,omitempty"`
 
-	// Content is the actual content of the file.
-	Content string `json:"content"`
+	// Content is the actual content of the file. Mutually exclusive with ContentFrom.
+	// +optional
+	Content string `json:"content,omitempty"`
+
+	// ContentFrom indicates the content for the file should be retrieved from a referenced resource. Mutually exclusive with Content.
+	// +optional
+	ContentFrom *FileContentSource `json:"contentFrom,omitempty"`
```

Comment:
Would it make sense to move this up a level? Instead of requiring one secret or configmap per file, leverage the structure of secrets and configmaps to allow a user to import multiple files from a single configmap or secret?

Reply:
That would mean we'd have `files` for inline, and `filesFrom` (?) for content from confimaps/secrets?

Reply:
> That would mean we'd have files for inline, and filesFrom (?) for content from confimaps/secrets?

Possibly...  There is also the issue of the additional metadata that is included in the individual file types as well. It just feels quite a bit heavy weight to force one secret/configmap per file, when these resources are intended to provide multiple files by design.

Reply:
Right, in light of this, I think we probably should spend some more time data modeling.

Reply:
Perhaps it makes sense to copy the same pattern that pod volumes use:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#configmapvolumesource-v1-core
Allow for specifying default file metadata (in this case `owner`, `permissions`, `encoding`) at the directory level (see `defaultMode` in the above example), and then allow for overriding those at the file level (see `items[].mode` in the above example).

Reply:
That might work. We could have filesFrom contain a *ConfigMapVolumeSource and a *SecretVolumeSource, or if that isn't sufficient due to lack of owner and encoding, we could make our own expanded types that have those fields.

Reply:
@vincepri any thoughts here?

Reply:
+1 on the pattern linked. I'd say if we go down that route, we should probably do our own expanded types, which makes sense in this case given that we might want to support more fields.

Reply:
Cross posting a [comment from the issue](https://github.com/kubernetes-sigs/cluster-api/issues/1846#issuecomment-564678587)

Please, while designing the API around this feature please consider also the need to support "Pivot/Move" of those resources. The part that most concerns me is to understand if more that one KubeadmConfig can reference to the same ConfigMap/Secret

The current Pivot logic moves one machine at time/with all the dependencies; having two KubeadmConfig references the same ConfigMap might break this process when I will be processing the second KubeadmConfig the ConfigMap will be already gone.
The case when the two KubeadmConfig belong to different clusters is even worst because clusters can now be migrated at different time

Reply:
Here is a possible data structure optimized for deduplication of Encoding, Permissions, and Owner.

The downsides that I see to this approach:
1. I am not sure there will be a strong correlation between Encoding/Perms/Owner and file source (configmap/secret/inline).
2. The original structure [more closely mapped to cloud config](https://cloudinit.readthedocs.io/en/latest/topics/examples.html#writing-out-arbitrary-files) `.write_files` structure which was slightly more intuitive to me.

Possible modifications:
A. Allow `.configMap` / `.secret` / `.content` to coexist in same `.fileSet[i]`.
B. Modify `.configMap` / `.secret` to be lists (i.e. `.configMaps` / `.secrets`).

Thoughts? Recommendations? @detiber @vincepri @ncdc 

```yaml
spec:
  fileSets:
  - defaultEncoding: ""
    defaultPermissions: ""
    defaultOwner: ""
    configMap:
      name: my-config-map
      items:
      - path: /tmp/foo.txt
        key: foo
        encoding: ""
        permissions: ""
        owner: ""
      ...
  - defaultEncoding: ""
    defaultPermissions: ""
    defaultOwner: ""
    secret:
      name: my-secret
      items:
      - path: /tmp/bar.txt
        key: bar
        encoding: ""
        permissions: ""
        owner: ""
      ...
  - defaultEncoding: ""
    defaultPermissions: ""
    defaultOwner: ""
    inline: # Inline content, possibly deprecate `.spec.files` in favor of this (.spec.fileSets[].inline[])?
    - path: /tmp/bar.txt
      content: "inline content goes here"
      encoding: ""
      permissions: ""
      owner: ""
    ...
     
     
```

Reply:
I think I like this with modification B

Reply:
Do we need to handle both configMap and secret?

Reply:
@ncdc Do you see downsides to A?

Reply:
> Do we need to handle both configMap and secret?

Seems reasonable to me to support ConfigMaps for teams that prefer to store non-sensitive files outside of Secrets. @vincepri Do you see a reason to avoid it?

Reply:
I'm +1 to both secrets & configmaps.

Oh, and yes, both A & B 😄

Reply:
I like the idea of option A, and have no objections to supporting both configmaps and secrets.

Reply:
Following up on the design (also with ConfigMaps lists), here would be a minimal config for specifying a single file.

Opinions on whether this appears sufficiently simple/intuitive to a user?

```yaml
spec:
  fileSets:
  - configMaps:
    - name: my-config-map
      items:
      - path: /tmp/foo.txt
        key: foo
```

Reply:
I'm wondering if we could do some defaulting around the key/path?
Maybe a pathPrefix at the configMap/secret level? then you could default items for each key/value pair in the configmap/secret.

Reply:
Do the configmap & secret sources in Kubernetes offer anything like this?

Reply:
> I'm wondering if we could do some defaulting around the key/path?
> Maybe a pathPrefix at the configMap/secret level? then you could default items for each key/value pair in the configmap/secret.

Makes sense to me... this would probably avoid the need for specifying path for every item.

Reply:
> Do the configmap & secret sources in Kubernetes offer anything like this?

Pod volumes are a slightly different use-case in that the unit of concern is a dir rather than a file. Also, the volume source and the volume mount location are specified separately.

Reply:
It should be documented in the code how we are preventing two machines referencing the same ConfigMap/Secrets, because this will break the clusterctl move logic.

Reply:
@fabriziopandini Noted.

Are we happy to move forward with the above data structure?

Reply:
I'm ok with it.
FYI I'm trying to overcome the limitation of the single cluster on a parallel thread about clusterctl move

Reply:
I'm ok moving forward. Note, we have a fairly aggressive timeline for getting v1alpha3 done, so this may make it, or be deferred to v1alpha4 - just want to be up front about that.